plugins/version-control-systems/git/src/main/kotlin/Git.kt

```diff
@@ -218,16 +220,22 @@ class Git(
 
             // See https://git-scm.com/docs/gitrevisions#_specifying_revisions for how Git resolves ambiguous
             // names. In particular, tag names have higher precedence than branch names.
-            runCatching {
-                fetch.setRefSpecs(revision).call()
-            }.recoverCatching {
-                // Note that in contrast to branches / heads, Git does not namespace tags per remote.
-                val tagRefSpec = "+${Constants.R_TAGS}$revision:${Constants.R_TAGS}$revision"
-                fetch.setRefSpecs(tagRefSpec).call()
-            }.recoverCatching {
-                val branchRefSpec = "+${Constants.R_HEADS}$revision:${Constants.R_REMOTES}origin/$revision"
-                fetch.setRefSpecs(branchRefSpec).call()
-            }.getOrThrow()
+            val refSpec = if (SHA1_REGEX.matches(revision)) {
+                revision
+            } else {
+                val refs = git.lsRemote().call()
```

Comment:
I guess this call makes things slower again. Can't we just "blindly" fetch tags and branches with

    fetch.setRefSpecs(branchRefSpec, tagRefSpec).call()

and only keep handling the SHA1 `revision` specially?

Reply:
No, the fetch call throws an exception if a provided ref spec does not exist and then the fallback to fetch all revision is invoked. This would only work in the rare situation where there is a tag and a branch with the same name.

Reply:
> No, the fetch call throws an exception if a provided ref spec does not exist

I wonder whether we can trick around that behavior by changing the refspecs to end with a `*`:

    "+${Constants.R_TAGS}$revision*:${Constants.R_TAGS}$revision*"

in some rare cases, this could fetch too much (if `revision` is a prefix of another revision), but still the code would check out the correct revision in the end.

Reply:
I need to test if JGit does not throw an exception if there are no refs matching these patterns.

Reply:
When I add the `*` the fetch does not fail, but the checkout fails with:
```
java.io.IOException: Running 'git checkout main' in '/home/user/tmp/test2' failed with exit code 1:
error: pathspec 'main' did not match any file(s) known to git
```

Reply:
> the checkout fails

Oh, that's very much unexpected. But hey 🤷🏻‍♂️